### DISCUSSION TITLE
Supabase not refreshing user auth token in middleware

### DISCUSSION THREAD

**Discussion by msaaddev** [supabase/supabase#28069] [Questions] [answered]
I have followed this SSR official [guide](https://supabase.com/docs/guides/auth/server-side/nextjs?router=app) to set up SSR Auth in my app. Now I am running into a strange issue. 

When the user resets his password, my `/confirm` route gets it and the `token_hash` gets verified successfully using `verifyOtp()`. There is no error and I see a response from the SDK just like it mentioned in the docs [here](https://supabase.com/docs/reference/javascript/auth-verifyotp). The user is then redirected to an authenticated page where he can change his password.

Between this route and `page.tsx`, the middleware gets hit and it should refresh the session but for some reason, it's not doing it. And I am getting the following response when I call `await supabase.auth.getUser()`:

```
user error  [AuthSessionMissingError: Auth session missing!] {
  __isAuthError: true,
  name: 'AuthSessionMissingError',
  status: 400,
  code: undefined
}
```

This is what my middleware code looks like:

```ts
import { createServerClient } from '@supabase/ssr';
import { NextResponse, type NextRequest } from 'next/server';

/**
 * This Middleware example can be used to refresh expired sessions before loading Server Component routes.
 * Server Components only have read access to cookies.
 *
 * Documentation: https://supabase.com/docs/guides/auth/server-side/nextjs?router=app
 *
 * @param {NextRequest} request - The NextRequest object representing the incoming request.
 * @returns {Promise<NextResponse>} The NextResponse object representing the outgoing response.
 */
export async function authMiddleware(request: NextRequest) {
	let supabaseResponse = NextResponse.next({
		request
	});

	const supabase = createServerClient(
		process.env.NEXT_PUBLIC_SUPABASE_URL!,
		process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
		{
			cookies: {
				getAll() {
					return request.cookies.getAll();
				},
				setAll(cookiesToSet) {
					cookiesToSet.forEach(({ name, value, options }) =>
						request.cookies.set(name, value)
					);
					supabaseResponse = NextResponse.next({
						request
					});
					cookiesToSet.forEach(({ name, value, options }) =>
						supabaseResponse.cookies.set(name, value, options)
					);
				}
			}
		}
	);

	const {
		data: { user },
		error
	} = await supabase.auth.getUser();

	console.log('get user', user);
	console.log('user error', error);

	return supabaseResponse;
}
```

It's exact as per the documentation. I have tried logging `request.cookies.getAll()` and there is a cookie for `sb-localhost-auth-token-code-verifier`

```
[
	{
		name: 'sb-localhost-auth-token-code-verifier',
		value: 'base64-IjEyNmU3MzdkYmJiZTllNGM5MzhhOGZlYjc1M2U3YThjZDBhZGM4MWI5NmJhYjQ4MmYxZTgxOTM3NGQzNzVhMDVjNWRkNmFiMTk0YTE0OTA2YmYwMjk5OWE0NWFhN2I5MGMzZTg4YjY2ZTdiYzM3NTMvUEFTU1dPUkRfUkVDT1ZFUlki'
	}
];
```

All this was previously working but now something is causing the session not to refresh in the middleware and I am not sure what. Will appreciate some help here! It's a major issue ATM.

More context:

1. 1st picture is from `confirm/route.ts` route. See that `auth confirm error` is null and we have `auth confirm data` that we get if the `token_hash` is successfully validated.
2. Second picture is from `supabase/middleware.ts` responsible for refreshing the session by calling `await supabase.auth.getUser()` (code provided above) but although request cookies exist, the user couldn't be fetched and session wasn't refreshed.

![screenshot-2024-07-18-at-06 18 47PM@2x](https://github.com/user-attachments/assets/557fb296-0a2a-4f96-847a-126203819426)
![screenshot-2024-07-18-at-06 19 02PM@2x](https://github.com/user-attachments/assets/39bc2ea9-5742-4f79-8038-34851ce47ef0)

**Comment by silentworks** [comment]
It's best to create a minimal reproducible example for someone to take a look at. I have an example NextJS setup here doing what you are asking about with no `Auth session missing` issue. https://github.com/silentworks/supabase-by-example/tree/main/nextjs

**Comment by msaaddev** [comment, marked as answer]
This was caused because `createServerClient` was not handling `PASSWORD_RECOVERY` event. I needed to update `@supabase/ssr` package to `0.4.1` as it contained the fix. PR [here](https://github.com/supabase/ssr/pull/34).

Thank you Supabase support for the help here!
